src/Microsoft.DotNet.Build.VSTSBuildsApi/VSTSBuildClient.cs

```diff
@@ -0,0 +1,228 @@
+        {
+            if(string.IsNullOrWhiteSpace(collectionIdentifier))
+            {
+                throw new Exception("Required parameter (collectionIdentifier) cannot be null or white space.");
+            }
+            if(!Uri.IsWellFormedUriString(collectionIdentifier, UriKind.Relative))
+            {
+                throw new Exception(string.Format("collectionIdentifier '{0}' contains invalid characters.  collectionIdentifier must contain only valid uri characters.", collectionIdentifier));
+            }
+
+            _collectionIdentifier = collectionIdentifier;
+            _credentials = credentials;
+        }
+
+        /// <summary>
+        /// Load a local build definition, combine the uniqueIdentifier with the definition name, 
+        /// check VSTS for a matching definition name, if present, update that definition with the
+        /// local build definition, otherwise create a new definition.
+        /// </summary>
+        /// <param name="stream">Stream to a VSTS build definition</param>
+        /// <returns>Created or updated build Id</returns>
+        public string CreateOrUpdateDefinition(Stream stream)
+        {
+            JObject definition = GetDefinition(stream);
+            return CreateOrUpdateDefinition(definition);
+        }
+
+        /// <summary>
+        /// Load a local build definition, combine the uniqueIdentifier with the definition name, 
+        /// check VSTS for a matching definition name, if present, update that definition with the
+        /// local build definition, otherwise create a new definition.
+        /// </summary>
+        /// <param name="definition">VSTS build definition JSON object</param>
+        /// <returns>Created or updated build definition id</returns>
+        private string CreateOrUpdateDefinition(JObject definition)
+        {
+            var key = string.Join("_", _collectionIdentifier, definition["name"].ToString());
+            definition["name"] = key;
+            var vstsDefinitions = VSTS_RetrieveDefinitionsListByNameAndPath(definition);
+
+
+            if (vstsDefinitions.Count == 0)
+            {
+                /* Create */
+                VSTS_CreateDefinition(definition);
+                definition = VSTS_RetrieveDefinitionById(definition);
+            }
+            else if(vstsDefinitions.Count == 1)
+            {
+                var vstsDefinition = vstsDefinitions[0];
+
+                /* Update */
+                if (!IsSameContents(definition, vstsDefinition))
+                {
+                    definition["revision"] = vstsDefinition["revision"];
+                    definition["id"] = vstsDefinition["id"];
+                    definition["uri"] = vstsDefinition["uri"];
+                    definition["path"] = vstsDefinition["path"];
+                    definition["createdDate"] = vstsDefinition["createdDate"];
+                    definition["url"] = vstsDefinition["url"];
```

Comment:
I'm wondering if it's worth pulling out the key names into arrays here and the other places in the file. It looks like they're a little different in each case, though.